collector/collector.go

```diff
@@ -0,0 +1,874 @@
+
+	policy.AddNewStringRule(configKey,
+		"procfs",
+		false,
+		plugin.SetDefaultString("/proc"))
+
+	return *policy, nil
+}
+
+type collector struct {
+	containers map[string]*container.ContainerData // holds data for a container under its short id
+	client     container.DockerClientInterface     // client for communication with docker (basic info, stats, mount points)
+	cgroupfs   string
+	driver     string
+	mounts     map[string]string
+	conf       map[string]string
+}
+
+// getRidGroup returns quested metrics grouped by docker ids
+func (c *collector) getRidGroup(mt ...plugin.Metric) (map[string]map[string]struct{}, error) {
+	ridGroup := make(map[string]map[string]struct{})
+	for _, m := range mt {
+		ns := m.Namespace.Strings()
+		if len(ns) < lengthOfNsPrefix+2 {
+			return nil, fmt.Errorf("Invalid name of metric %+s", strings.Join(ns, "/"))
+		}
+
+		rid := ns[2]
+
+		group, err := getQueryGroup(ns[3:])
+		if err != nil {
+			return nil, err
+		}
+
+		switch rid {
+		case "*":
+			for id := range c.containers {
+				appendIfMissing(ridGroup, id, group)
+			}
+		case "root":
+			appendIfMissing(ridGroup, "root", group)
+		default:
+			shortID, err := container.GetShortID(rid)
+			if err != nil {
+				return nil, err
+			}
+
+			if _, exist := c.containers[shortID]; !exist {
+				return nil, fmt.Errorf("Docker container %+s cannot be found", rid)
+			}
+
+			appendIfMissing(ridGroup, shortID, group)
+		}
+	}
+
+	if len(ridGroup) == 0 {
+		return nil, fmt.Errorf("can't retrieve docker ids and requestd metrics")
+	}
+
+	return ridGroup, nil
+}
+
+func (c *collector) collect(ridGroup map[string]map[string]struct{}, procfs string) error {
+	var err error
+	for rid, groups := range ridGroup {
+		opts := make(container.GetStatOpt)
+		opts["procfs"] = procfs
+
+		if rid == "root" {
+			opts["is_host"] = true
+			opts["pid"] = -1
+			opts["container_id"] = "root"
+			opts["container_drv"] = c.driver
+		} else {
+			cont, err := c.client.InspectContainer(rid)
+			if err != nil {
+				return err
+			}
+			opts["is_host"] = false
+			opts["pid"] = cont.State.Pid
+			opts["container_id"] = cont.ID
+			opts["container_drv"] = cont.Driver
+		}
+
+		for group := range groups {
+			// during initialization of docker client information about running containers is collectd
+			if group == "spec" {
+				continue
+			}
+
+			if group == "pids_stats" && rid == "root" {
+				continue
+			}
+
+			if group != "network" && group != "tcp" && group != "tcp6" && group != "filesystem" {
+				cgroup := names[group]
+				// try to find cgroup mount point in cache
+				cpath, exists := c.mounts[cgroup]
+				if !exists {
+					cpath, err = c.client.FindCgroupMountpoint(procfs, cgroup)
+					if err != nil {
+						return err
+					}
+					c.mounts[cgroup] = cpath
+				}
+
+				if rid != "root" {
+					if c.cgroupfs == "systemd" {
+						cpath = filepath.Join(cpath, "system.slice", fmt.Sprintf("docker-%s.scope", c.containers[rid].ID))
```

Comment:
What do you think about adding here a log with information about the path to cgroups? I suppose it will be helfpul based on issues which were reported in the past.